psl/psl-core/src/validate/validation_pipeline/validations/relation_indexes.rs

```diff
@@ -0,0 +1,78 @@
+use diagnostics::DatamodelWarning;
+use parser_database::{
+    ast::{FieldId, ModelId},
+    walkers::{RelationFieldWalker, Walker},
+};
+
+use crate::{datamodel_connector::RelationMode, validate::validation_pipeline::context::Context};
+
+// { x_1, x_2, ..., x_n } is left-wise included in { y_1, y_2, ..., y_m } if and only if
+// n <= m and x_i = y_i for all i in [1, n].
+fn is_left_wise_included<T>(item: &[T], group: &[T]) -> bool
+where
+    T: PartialEq,
+{
+    group.iter().take(item.len()).eq(item.iter())
+}
+
+pub(super) fn validate_missing_relation_indexes(
+    model: Walker<'_, ModelId>,
+    relation_field: RelationFieldWalker<'_>,
+    ctx: &mut Context<'_>,
+) {
+    let is_provider_mongodb = ctx
+        .datasource
+        .map(|datasource| datasource.active_provider == "mongodb")
+        .unwrap_or(false);
+
+    if is_provider_mongodb || ctx.relation_mode != RelationMode::Prisma {
+        return;
+    }
+
+    if let Some(fields) = relation_field.referencing_fields() {
+        // Collects all fields that should be part of an index in the given model, w.r.t. to left-wise inclusion.
+        let relation_fields: Vec<FieldId> = fields.map(|field| field.field_id()).collect();
+
+        // Collects all groups of indexes explicitly declared in the given model.
+        // An index group can be:
+        // - a singleton (@unique or @id)
+        // - an ordered set (@@unique or @@index)
+        let index_sets: Vec<Vec<FieldId>> = model
+            .indexes()
+            .map(|index_walker| index_walker.fields().map(|index| index.field_id()).collect())
+            .collect();
+
+        let relation_fields_appear_in_index = index_sets
+            .iter()
+            .any(|index_set| is_left_wise_included(&relation_fields, index_set));
```

Comment:
The whole code from line 32 to here can be simplified to

```

let relation_fields_appear_in_index = model.indexes().any(|idx| {
   let fields = fields.clone();
   let index_fields = index.fields();
   fields.len() <= index_fields.len() && fields.zip(index_fields).all(|(a, b)| a.field_id == b.field_id)
})
```

It's shorter and avoids all the allocations.

Reply:
(it's pseudo-code but I'm confident it's very close to compiling)

Reply:
Ok, taking a look at this. `fields` cannot be cloned:

```
expected `impl ExactSizeIterator + Iterator<Item = parser_database::walkers::ScalarFieldWalker<'_>>` to be an iterator that yields `&_`, but it yields `parser_database::walkers::ScalarFieldWalker<'_>`
expected reference `&_`
      found struct `parser_database::walkers::ScalarFieldWalker<'_>`
```

Reply:
The only way I've found it to collect `fields` at least once:

```rust
        let relation_fields: Vec<_> = fields.collect();

        let relation_fields_appear_in_index = model.indexes().any(|index_walker| {
            let fields = &relation_fields;
            let index_fields = index_walker.fields();
            relation_fields.len() <= index_fields.len()
                && fields
                    .iter()
                    .zip(index_fields)
                    .all(|(a, b)| a.field_id() == b.field_id())
        });
```

Reply:
(This `any + (zip + all)` logic seems wrong, I see tests failing. Reasoning about this)

Reply:
> Ok, taking a look at this. fields cannot be cloned, there's a reference mismatch:

You can add a `+ Clone` to the return type of `fields()` where it is defined.

Reply:
(Note that the above was wrong and made tests fail).
I have replaced that with:

```rust
let referencing_fields: Vec<_> = fields.map(|field| field.field_id()).collect();

let referencing_fields_appear_in_index = model
    .indexes()
    .map(|index_walker| index_walker.fields().map(|index| index.field_id()))
    .any(|index_fields_it| {
        let is_leftwise_included_new = referencing_fields.len() <= index_fields_it.len()
            && referencing_fields
                .iter()
                .zip(index_fields_it)
                .all(|(referencing_field, index_field)| *referencing_field == index_field);

        is_leftwise_included_new
    });
```